Fix custom columns lose their definition in embeddings

WiNloSt opened the pull request:
Fixes: #25473
[Notion summary](https://www.notion.so/metabase/Question-about-custom-field-definition-in-parameters-on-public-dashboards-641c1603918548a2a1bc18fd430fd901)

## PR Summary
In `<FieldValuesWidget />`, there could be the `parameter` prop passed to it which indicates the widget is connected to a dashboard filter.

In such case, if we need to perform any logic to determine the widget type e.g. showing the placeholder as `Enter a number` we should use `parameter`'s type instead of field types.

## Todos
- [x] Add tests
- [x] Troubleshoot `ExpressionDimension.parseMBQL`
- [x] Draft questions about new public table metadata endpoint

<details>
<summary><h2>Old details</h2></summary>

## Problems
When populating parameters we use this selector:
https://github.com/metabase/metabase/blob/00ce5700e31a98124b034e0586456e9ce261505c/frontend/src/metabase/dashboard/selectors.js#L171-L179

Which down the line, uses the `Field` selector:
https://github.com/metabase/metabase/blob/00ce5700e31a98124b034e0586456e9ce261505c/frontend/src/metabase/selectors/metadata.js#L32

The problem is that we don't populate `state.entities.fields` in embedding.

In the app we call `/api/table/:tableId/query_metadata` here:
https://github.com/metabase/metabase/blob/00ce5700e31a98124b034e0586456e9ce261505c/frontend/src/metabase/entities/tables.js#L95-L98
which will populate `state.entities.fields` as expected, but in embedding we only call either `/api/public/dashboard/:uuid/dashcard/:dashcardId/card/:cardId` or `/api/embed/dashboard/:token/dashcard/:dashcardId/card/:cardId`:
https://github.com/metabase/metabase/blob/00ce5700e31a98124b034e0586456e9ce261505c/frontend/src/metabase/dashboard/actions/data-fetching.js#L378
which only populate `dashboard.dashcards.[cardId]`

I think of a couple solutions
1. Modify the selector so that it pulls fields from `dashboard.dashcards.[cardId]`
2. Kind of what this PR does, is to implement the reducer to populate `state.entities.fields`, but I found some problems which are laid out in the comments.
</details>

WiNloSt commented on frontend/src/metabase/entities/fields.js:
```diff
@@ -191,6 +191,14 @@ const Fields = createEntity({
           Array.from(new Map(existing.concat(remappings))),
         ),
+      // Can't import metabase/dashboard/FETCH_CARD_DATA because of circular dependency
+      ["metabase/dashboard/FETCH_CARD_DATA"]: (state, { payload }) => {
```
Use hard-coded action since `metabase/dashboard/FETCH_CARD_DATA` can't be imported due to cyclic dependencies.

WiNloSt commented on frontend/src/metabase/entities/fields.js:
```diff
@@ -191,6 +191,14 @@ const Fields = createEntity({
+      ["metabase/dashboard/FETCH_CARD_DATA"]: (state, { payload }) => {
+        const fields = getIn(payload, ["result", "data", "cols"]);
+        return {
+          ...Fields.normalizeList(fields).entities[Fields.name],
```
I really hate to have to normalize the data manually. I bet this isn't idiomatic, but I don't know how to get the data in the correct shape without normalizing it.

Under the hood, calling `normalizeList` will populate `uniqueId` which will be used by some of our selectors.
https://github.com/metabase/metabase/blob/00ce5700e31a98124b034e0586456e9ce261505c/frontend/src/metabase/selectors/metadata.js#L128

without that, `fields` won't be populated using the `getMetadata` selector.

kulyk replied:
Let me take another look at it though

ranquild reviewed (commented):
@WiNloSt I'd ask for a review from @kulyk but based on my limited understanding `data.cols` is not the same as fields obtained from metadata endpoints. What would happen if we allow this request to be done for embedded dashboards, too https://github.com/metabase/metabase/blob/fa76af8cf88eeb44a6c5581ea8ac18a30b75ebf6/frontend/src/metabase/dashboard/actions/data-fetching.js#L177? If that won't work, I believe we should start a conversation with the backend team whether they can make the existing endpoint work or add another one that can be used for embedding.

WiNloSt committed:
Fix input widget with parameter using incorrect type

When a FieldValuesWidget as `parameter` prop, it indicates that it is
tied to a dashboard parameter, so we should use parameter instead of
field type when rendering the component.

WiNloSt commented on frontend/src/metabase-lib/parameters/utils/parameter-type.ts:
```diff
@@ -22,6 +22,11 @@ function splitType(parameterOrType: Parameter | string) {
   return parameterType.split("/");
 }
 
+export function isIdParameter(parameter: Parameter | string) {
```
Keep the logic to check parameter type in here. I think the signature could be refactored to `isParameterType(parameter: Parameter, type: ParameterType)`, so we don't have to import separate function to determine parameter types.

WiNloSt commented on frontend/src/metabase/components/FieldValuesWidget.jsx:
```diff
@@ -291,6 +296,7 @@ class FieldValuesWidgetInner extends Component {
 
     const tokenFieldPlaceholder = getTokenFieldPlaceholder({
       fields,
+      parameter,
```
This is the first place where we didn't take parameter into account, basically when there's a dashboard filter (parameter) set, we should use that filter's type instead of field types.

WiNloSt commented on frontend/src/metabase/components/FieldValuesWidget.jsx:
```diff
@@ -307,6 +313,12 @@ class FieldValuesWidgetInner extends Component {
     const hasListValues = hasList({ fields, disableSearch, options });
 
+    const parseFreeformValue = value => {
+      return isNumeric(fields[0], parameter)
```
This is the second place, where we didn't take parameter type into account.

WiNloSt commented on frontend/src/metabase/components/FieldValuesWidget.jsx:
```diff
@@ -446,17 +454,30 @@ function shouldList(fields, disableSearch) {
-    return t`Enter an ID`;
-  } else if (firstField.isString()) {
+function getNonSearchableTokenFieldPlaceholder(firstField, parameter) {
+  if (parameter) {
```
I despise nested if, but I just want to keep things simple for now to verify if the approach is valid first.

ranquild commented on frontend/src/metabase/components/FieldValuesWidget.jsx:
```diff
@@ -671,3 +693,17 @@ export function getValuesMode(
+/**
+ *
+ * @param {import("metabase-types/types/Field").Field} field
+ * @param {import("metabase-types/types/Parameter").Parameter} parameter
```
Types are wrong. `field` is coming from `metabase-lib` (it's a class), and parameter is `UiParameter` from `metabase-lib/parameters/types`. But I'd remove this comment because it will become outdated very shortly.

WiNloSt replied:
I was using the same type from `isNumberParameter`'s function signature

https://github.com/metabase/metabase/blob/2bd929c6fc1e7a0d0077c7cee903f9dd2ad817da/frontend/src/metabase-lib/parameters/utils/parameter-type.ts#L35

which seems to receive the same `parameter` we have here.
https://github.com/metabase/metabase/blob/2bd929c6fc1e7a0d0077c7cee903f9dd2ad817da/frontend/src/metabase/parameters/components/ParameterValueWidget.jsx#L234

We seem to have types in a lot of places, so that's why I was confused.

But agree on removing this comment since it'll be refactored soon.

ranquild reviewed (changes requested):
@WiNloSt I believe that the PR does parameter checks correctly. But it's hacky because we overload `FieldValuesWidget` with a notion of `parameter` instead of creating a base component that would work both for `FieldValuesWidget` and for parameters. But I'd approve the PR if there would be tests for the behavior.

WiNloSt committed:
Add a test

WiNloSt commented:
@ranquild OK if this approach sounds good I can refactor the PR tomorrow. BTW the test has just been added. I verify that without the fix the test would fail, so it's capturing the bug correctly.

ranquild reviewed (commented):
@WiNloSt I'm ok with refactoring this in a separate PR. If the fix works this is good already.

WiNloSt commented on frontend/test/metabase/scenarios/sharing/public.cy.spec.js:
```diff
@@ -201,4 +201,106 @@ describe("scenarios > public", () => {
     );
   });
+
+  describe("public dashboard with parameters linked to a custom field (metabase#25473)", () => {
```
Mostly copy the setup from https://github.com/metabase/metabase/blob/master/frontend/test/metabase/scenarios/embedding/embedding-dashboard.cy.spec.js

ranquild commented on frontend/test/metabase/scenarios/sharing/public.cy.spec.js:
```diff
@@ -201,4 +201,106 @@ describe("scenarios > public", () => {
+      // Enable public sharing
+      cy.icon("share").click();
+
+      cy.contains("Enable sharing")
```
nit: I think we usually prefer `cy.findByText`

ranquild commented on frontend/test/metabase/scenarios/sharing/public.cy.spec.js:
```diff
@@ -201,4 +201,106 @@ describe("scenarios > public", () => {
+      // Public dashboard
+      cy.findByText(textContainsFilter.name).click();
+
+      cy.findByPlaceholderText("Enter some text").should("be.visible");
```
We should also check that it's possible to enter a value and it doesn't break

WiNloSt replied:
sounds good  👍

ranquild reviewed (changes requested):
Please add another case to the test 👍

WiNloSt committed:
Address review

WiNloSt commented on frontend/test/metabase/scenarios/sharing/public.cy.spec.js:
```diff
@@ -201,4 +201,130 @@ describe("scenarios > public", () => {
+          cy.visit(dashboardPublicLink);
+        });
+
+      // Assert public dashboard
```
Given the custom field as `concat([Email], [Name])`
There are 3 assertions here.
1. search by email
2. search by name
3. search by email+name

Each with different person, so we know it's working.

WiNloSt committed:
Remove wrong JSdoc types

ranquild commented:
@WiNloSt should we backport the PR as well?

WiNloSt commented:
@ranquild Good point. I think we should. This only fixes the bug. So no default behavior change and whatnot.

WiNloSt committed:
Fix failed E2E test due to wrong mock parameter type

WiNloSt commented on frontend/test/metabase/scenarios/filters/view.cy.spec.js:
```diff
@@ -37,15 +37,15 @@ describe("scenarios > question > view", () => {
               type: "dimension",
               dimension: ["field", PRODUCTS.CATEGORY, null],
-              "widget-type": "id",
+              "widget-type": "string/=",
```
@nemanjaglumac Can you help take a look that this change is fine? I need to correct the widget type to match the actual field type so `<FieldValuesWidget />` show the correct widget type.

The [recent E2E test failure](https://www.deploysentinel.com/ci/runs/63a32c02a622c2ef12225f15) is because we mock the widget type wrong.

Since both **product.category** and **product.vendor** are both string, when we try to map the parameter using the UI, the type will be defaulted to `string/=`
<img width="330" alt="image" src="https://user-images.githubusercontent.com/1937582/209128119-003262fe-6c8d-47b6-b896-898658562a6c.png">

And with this new change that takes parameter type into account, we need to make sure we mock the correct parameter type.

nemanjaglumac replied:
I don't remember exactly what was going on in this test, but your change makes sense.

Thank you
